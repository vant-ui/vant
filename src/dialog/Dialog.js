import { reactive } from 'vue';

// Utils
import { callInterceptor } from '../utils/interceptor';
import { createNamespace, addUnit, pick } from '../utils';
import { BORDER_TOP, BORDER_LEFT } from '../utils/constant';

// Components
import Popup, { popupSharedProps } from '../popup';
import Button from '../button';
import ActionBar from '../action-bar';
import ActionBarButton from '../action-bar-button';

const [createComponent, bem, t] = createNamespace('dialog');

const popupKeys = [
  ...Object.keys(popupSharedProps),
  'transition',
  'closeOnPopstate',
];

export default createComponent({
  props: {
    ...popupSharedProps,
    title: String,
    theme: String,
    width: [Number, String],
    message: String,
    callback: Function,
    allowHtml: Boolean,
    className: null,
    beforeClose: Function,
    messageAlign: String,
    showCancelButton: Boolean,
    cancelButtonText: String,
    cancelButtonColor: String,
    confirmButtonText: String,
    confirmButtonColor: String,
    closeOnClickOverlay: Boolean,
    transition: {
      type: String,
      default: 'van-dialog-bounce',
    },
    showConfirmButton: {
      type: Boolean,
      default: true,
    },
    closeOnPopstate: {
      type: Boolean,
      default: true,
    },
  },

  emits: ['confirm', 'cancel', 'update:show'],

  setup(props, { emit, slots }) {
    const loading = reactive({
      confirm: false,
      cancel: false,
    });

    const onUpdateShow = (value) => {
      emit('update:show', value);
    };

    const close = (action) => {
      onUpdateShow(false);
      if (props.callback) {
        props.callback(action);
      }
    };

    const handleAction = (action) => {
      // should not trigger close event when hidden
      if (!props.show) {
        return;
      }

      emit(action);

      if (props.beforeClose) {
        loading[action] = true;
        callInterceptor({
          interceptor: props.beforeClose,
          args: [action],
          done() {
            close(action);
            loading[action] = false;
          },
          canceled() {
            loading[action] = false;
          },
        });
      } else {
        close(action);
      }
    };

    const renderTitle = () => {
      const title = slots.title ? slots.title() : props.title;
      if (title) {
        return (
          <div
            class={bem('header', {
              isolated: !props.message && !slots.default,
            })}
          >
            {title}
          </div>
        );
      }
    };

    const renderContent = () => {
      if (slots.default) {
        return <div class={bem('content')}>{slots.default()}</div>;
      }

      const { title, message, allowHtml, messageAlign } = props;
      if (message) {
        const hasTitle = title || slots.title;
        return (
          <div
            // add key to force re-render
            // see: https://github.com/youzan/vant/issues/7963
            key={allowHtml ? 1 : 0}
            class={bem('content', { isolated: !hasTitle })}
          >
            <div
              class={bem('message', {
                'has-title': hasTitle,
                [messageAlign]: messageAlign,
              })}
              {...{
                [allowHtml ? 'innerHTML' : 'textContent']: message,
              }}
            />
          </div>
        );
      }
    };

    const renderButtons = () => (
      <div class={[BORDER_TOP, bem('footer')]}>
        {props.showCancelButton && (
          <Button
            size="large"
            text={props.cancelButtonText || t('cancel')}
            class={bem('cancel')}
            style={{ color: props.cancelButtonColor }}
            loading={loading.cancel}
            onClick={() => {
              handleAction('cancel');
            }}
          />
        )}
        {props.showConfirmButton && (
          <Button
            size="large"
            text={props.confirmButtonText || t('confirm')}
            class={[bem('confirm'), { [BORDER_LEFT]: props.showCancelButton }]}
            style={{ color: props.confirmButtonColor }}
            loading={loading.confirm}
            onClick={() => {
              handleAction('confirm');
            }}
          />
        )}
      </div>
    );

    const renderRoundButtons = () => (
      <ActionBar class={bem('footer')}>
        {props.showCancelButton && (
          <ActionBarButton
            size="large"
            type="warning"
            text={props.cancelButtonText || t('cancel')}
            class={bem('cancel')}
            color={props.cancelButtonColor}
            loading={loading.cancel}
            onClick={() => {
              handleAction('cancel');
            }}
          />
        )}
        {props.showConfirmButton && (
          <ActionBarButton
            size="large"
            type="danger"
            text={props.confirmButtonText || t('confirm')}
            class={bem('confirm')}
            color={props.confirmButtonColor}
            loading={loading.confirm}
            onClick={() => {
              handleAction('confirm');
            }}
          />
        )}
      </ActionBar>
    );

    return () => {
      const { width, title, theme, message, className } = props;
      return (
        <Popup
          role="dialog"
          class={[bem([theme]), className]}
          style={{ width: addUnit(width) }}
          aria-labelledby={title || message}
          {...{
            ...pick(props, popupKeys),
            'onUpdate:show': onUpdateShow,
          }}
        >
          {renderTitle()}
          {renderContent()}
          {theme === 'round-button' ? renderRoundButtons() : renderButtons()}
        </Popup>
      );
    };
  },
});
