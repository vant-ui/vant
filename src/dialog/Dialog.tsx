import { PropType, reactive, defineComponent } from 'vue';

// Utils
import { callInterceptor, Interceptor } from '../utils/interceptor';
import { createNamespace, addUnit, pick, UnknownProp } from '../utils';
import { BORDER_TOP, BORDER_LEFT } from '../utils/constant';

// Components
import Popup from '../popup';
import Button from '../button';
import ActionBar from '../action-bar';
import ActionBarButton from '../action-bar-button';
import { popupSharedProps, popupSharedPropKeys } from '../popup/shared';

const [name, bem, t] = createNamespace('dialog');

export type DialogTheme = 'default' | 'round-button';
export type DialogAction = 'confirm' | 'cancel';
export type DialogMessageAlign = 'left' | 'center' | 'right';

const popupKeys = [
  ...popupSharedPropKeys,
  'transition',
  'closeOnPopstate',
] as const;

export default defineComponent({
  name,

  props: {
    ...popupSharedProps,
    title: String,
    theme: String as PropType<DialogTheme>,
    width: [Number, String],
    message: String,
    callback: Function as PropType<(action?: DialogAction) => void>,
    allowHtml: Boolean,
    className: UnknownProp,
    beforeClose: Function as PropType<Interceptor>,
    messageAlign: String as PropType<DialogMessageAlign>,
    showCancelButton: Boolean,
    cancelButtonText: String,
    cancelButtonColor: String,
    confirmButtonText: String,
    confirmButtonColor: String,
    closeOnClickOverlay: Boolean,
    transition: {
      type: String,
      default: 'van-dialog-bounce',
    },
    showConfirmButton: {
      type: Boolean,
      default: true,
    },
    closeOnPopstate: {
      type: Boolean,
      default: true,
    },
  },

  emits: ['confirm', 'cancel', 'update:show'],

  setup(props, { emit, slots }) {
    const loading = reactive({
      confirm: false,
      cancel: false,
    });

    const updateShow = (value: boolean) => emit('update:show', value);

    const close = (action: DialogAction) => {
      updateShow(false);
      if (props.callback) {
        props.callback(action);
      }
    };

    const getActionHandler = (action: DialogAction) => () => {
      // should not trigger close event when hidden
      if (!props.show) {
        return;
      }

      emit(action);

      if (props.beforeClose) {
        loading[action] = true;
        callInterceptor({
          interceptor: props.beforeClose,
          args: [action],
          done() {
            close(action);
            loading[action] = false;
          },
          canceled() {
            loading[action] = false;
          },
        });
      } else {
        close(action);
      }
    };

    const onCancel = getActionHandler('cancel');
    const onConfirm = getActionHandler('confirm');

    const renderTitle = () => {
      const title = slots.title ? slots.title() : props.title;
      if (title) {
        return (
          <div
            class={bem('header', {
              isolated: !props.message && !slots.default,
            })}
          >
            {title}
          </div>
        );
      }
    };

    const renderContent = () => {
      if (slots.default) {
        return <div class={bem('content')}>{slots.default()}</div>;
      }

      const { title, message, allowHtml, messageAlign } = props;
      if (message) {
        const hasTitle = title || slots.title;

        return (
          <div
            // add key to force re-render
            // see: https://github.com/youzan/vant/issues/7963
            key={allowHtml ? 1 : 0}
            class={bem('content', { isolated: !hasTitle })}
          >
            <div
              class={bem('message', {
                'has-title': hasTitle,
                [messageAlign as string]: messageAlign,
              })}
              {...{
                [allowHtml ? 'innerHTML' : 'textContent']: message,
              }}
            />
          </div>
        );
      }
    };

    const renderButtons = () => (
      <div class={[BORDER_TOP, bem('footer')]}>
        {props.showCancelButton && (
          <Button
            size="large"
            text={props.cancelButtonText || t('cancel')}
            class={bem('cancel')}
            style={{ color: props.cancelButtonColor }}
            loading={loading.cancel}
            onClick={onCancel}
          />
        )}
        {props.showConfirmButton && (
          <Button
            size="large"
            text={props.confirmButtonText || t('confirm')}
            class={[bem('confirm'), { [BORDER_LEFT]: props.showCancelButton }]}
            style={{ color: props.confirmButtonColor }}
            loading={loading.confirm}
            onClick={onConfirm}
          />
        )}
      </div>
    );

    const renderRoundButtons = () => (
      <ActionBar class={bem('footer')}>
        {props.showCancelButton && (
          <ActionBarButton
            type="warning"
            text={props.cancelButtonText || t('cancel')}
            class={bem('cancel')}
            color={props.cancelButtonColor}
            loading={loading.cancel}
            onClick={onCancel}
          />
        )}
        {props.showConfirmButton && (
          <ActionBarButton
            type="danger"
            text={props.confirmButtonText || t('confirm')}
            class={bem('confirm')}
            color={props.confirmButtonColor}
            loading={loading.confirm}
            onClick={onConfirm}
          />
        )}
      </ActionBar>
    );

    return () => {
      const { width, title, theme, message, className } = props;
      return (
        <Popup
          role="dialog"
          class={[bem([theme]), className]}
          style={{ width: addUnit(width) }}
          aria-labelledby={title || message}
          {...{
            ...pick(props, popupKeys),
            'onUpdate:show': updateShow,
          }}
        >
          {renderTitle()}
          {renderContent()}
          {theme === 'round-button' ? renderRoundButtons() : renderButtons()}
        </Popup>
      );
    };
  },
});
