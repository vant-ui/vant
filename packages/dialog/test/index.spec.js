import Dialog from '../';
import demoTest from '../../../test/demo-test';

demoTest(Dialog);
