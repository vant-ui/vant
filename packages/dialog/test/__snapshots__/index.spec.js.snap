// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Dialog correctly 1`] = `
<section class="van-doc-demo-section demo-Dialog" style="background:undefined;">
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">消息提示</h2>
    <button class="van-button van-button--default van-button--normal">
      <!----><span class="van-button__text">Alert</span></button>
    <button class="van-button van-button--default van-button--normal">
      <!----><span class="van-button__text">无标题 Alert</span></button>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">消息确认</h2>
    <button class="van-button van-button--default van-button--normal">
      <!----><span class="van-button__text">Confirm</span></button>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">高级用法</h2>
    <button class="van-button van-button--default van-button--normal">
      <!----><span class="van-button__text">高级用法</span></button>
    <div class="van-dialog" style="display:none;">
      <!---->
      <div class="van-hairline van-dialog__content">
        <div placeholder="请输入用户名" class="van-cell van-hairline van-field">
          <!---->
          <div class="van-cell__title"><span>用户名</span>
            <!---->
          </div>
          <div class="van-cell__value">
            <input type="text" placeholder="请输入用户名" value="" class="van-field__control">
            <!---->
            <!---->
            <!---->
          </div>
          <!---->
        </div>
        <div placeholder="请输入密码" class="van-cell van-hairline van-field">
          <!---->
          <div class="van-cell__title"><span>密码</span>
            <!---->
          </div>
          <div class="van-cell__value">
            <input type="password" placeholder="请输入密码" value="" class="van-field__control">
            <!---->
            <!---->
            <!---->
          </div>
          <!---->
        </div>
      </div>
      <div class="van-dialog__footer van-dialog__footer--buttons">
        <button class="van-button van-button--default van-button--large van-dialog__cancel">
          <!----><span class="van-button__text">
        取消
      </span></button>
        <button class="van-button van-button--default van-button--large van-dialog__confirm van-hairline--left">
          <!----><span class="van-button__text">
        确认
      </span></button>
      </div>
    </div>
  </section>
</section>
`;
